### PR TITLE
binding/c: autogen request/spawn api

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Auto generate request and spawn C bindings.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

File: src/mpi/request/request_impl.c [hunk 1]
```diff
@@ -5,12 +5,7 @@
...
 
-#if !defined(MPIR_REQUEST_PTR_ARRAY_SIZE)
-/* use a larger default size of 64 in order to enhance SQMR performance */
-#define MPIR_REQUEST_PTR_ARRAY_SIZE 64
-#endif
-
-int MPIR_Cancel(MPIR_Request * request_ptr)
+int MPIR_Cancel_impl(MPIR_Request * request_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
```
Comment (minsii) [inline src/mpi/request/request_impl.c:10]: Can you separate the request related refactoring into a separate commit, likely before the autogeneration commit?

File: src/mpi/request/request_impl.c [hunk 2]
```diff
@@ -97,196 +92,151 @@ int MPIR_Cancel(MPIR_Request * request_ptr)
   fn_exit:
     return mpi_errno;
   fn_fail:
-
     goto fn_exit;
```
Comment (minsii) [inline src/mpi/request/request_impl.c:95]: Here is a empty line change.

File: src/mpi/request/request_impl.c [hunk 3]
```diff
@@ -97,196 +92,151 @@ int MPIR_Cancel(MPIR_Request * request_ptr)
...
-static int MPIR_Grequest_registered_finalizer = 0;
-static MPIR_Grequest_class *MPIR_Grequest_class_list = NULL;
-
-/* Any internal routines can go here.  Make them static if possible.  If they
-   are used by both the MPI and PMPI versions, use PMPI_LOCAL instead of
-   static; this macro expands into "static" if weak symbols are supported and
-   into nothing otherwise. */
-static int MPIR_Grequest_free_classes_on_finalize(void *extra_data ATTRIBUTE((unused)))
+int MPIR_Request_free_impl(MPIR_Request * request_ptr)
 {
```
Comment (minsii) [inline src/mpi/request/request_impl.c:99]: Can you please separate grequest commit also into a separate commit and please explain the reason?

File: src/include/mpir_misc.h [hunk 4]
```diff
@@ -91,11 +91,6 @@ int MPIR_Get_intranode_rank(MPIR_Comm * comm_ptr, int r);
 int MPIR_Close_port_impl(const char *port_name);
 int MPIR_Open_port_impl(MPIR_Info * info_ptr, char *port_name);
 
-int MPIR_Comm_join(int fd, MPI_Comm * intercomm);
-int MPIR_Comm_disconnect(MPIR_Comm * comm_ptr);
-int MPIR_Publish_name(const char *service_name, MPIR_Info * info_ptr, const char *port_name);
-int MPIR_Unpublish_name(const char *service_name, MPIR_Info * info_ptr, const char *port_name);
-
 /* Default routines for asynchronous progress thread */
```
Comment (minsii) [inline src/include/mpir_misc.h:94]: Why do we remove these MPIR_ functions?

File: src/mpi/spawn/spawn_impl.c [hunk 5]
```diff
@@ -78,7 +78,7 @@ static int MPIR_fd_recv(int fd, void *buffer, int length)
     return 0;
 }
 
-int MPIR_Comm_join(int fd, MPI_Comm * intercomm)
+int MPIR_Comm_join_impl(int fd, MPIR_Comm ** p_intercomm_ptr)
 {
```
Comment (minsii) [inline src/mpi/spawn/spawn_impl.c:82]: Can  you please move the refactoring of `MPIR_Comm_join` into a separate commit?

File: src/mpi/spawn/spawn_impl.c [hunk 6]
```diff
@@ -122,7 +122,7 @@ int MPIR_Comm_join(int fd, MPI_Comm * intercomm)
     mpi_errno = MPIR_Close_port_impl(local_port);
     MPIR_ERR_CHECK(mpi_errno);
 
-    MPIR_OBJ_PUBLISH_HANDLE(*intercomm, intercomm_ptr->handle);
+    *p_intercomm_ptr = intercomm_ptr;
```
Comment (minsii) [inline src/mpi/spawn/spawn_impl.c:126]: Why this change? Should we return the handle of `intercomm`?

File: test/mpi/errors/spawn/connect_timeout.c [hunk 7]
```diff
@@ -102,13 +102,21 @@ int test_mismatched_accept(MPI_Comm intra_comm, int gid)
 
         IF_VERBOSE(intra_rank == 0, ("client: connecting to <%s> with default timeout.\n", port));
         mpi_errno1 = MPI_Comm_connect(port, MPI_INFO_NULL, 0, intra_comm, &comm);
+        if (comm != MPI_COMM_NULL) {
+            IF_VERBOSE(intra_rank == 0, ("connection matched, freeing communicator.\n"));
+            MPI_Comm_free(&comm);
+        }
         if (mpi_errno1 != MPI_SUCCESS) {
```
Comment (minsii) [inline test/mpi/errors/spawn/connect_timeout.c:109]: It looks like the code change does not match the commit message. Here you assume that as long as `comm` is changed, it will be a valid comm. Should we free `comm` only when `mpi_errno1` is `MPI_SUCCESS`?

File: test/mpi/errors/spawn/connect_timeout.c [hunk 8]
```diff
@@ -118,11 +126,11 @@ int test_mismatched_accept(MPI_Comm intra_comm, int gid)
...
-    }
+        if (comm != MPI_COMM_NULL) {
+            IF_VERBOSE(intra_rank == 0, ("connection matched, freeing communicator.\n"));
+            MPI_Comm_free(&comm);
+        }
 
-    if (comm != MPI_COMM_NULL) {
-        IF_VERBOSE(intra_rank == 0, ("connection matched, freeing communicator.\n"));
-        MPI_Comm_free(&comm);
     }
```
Comment (minsii) [inline test/mpi/errors/spawn/connect_timeout.c:134]: I am a bit confused about this change. It looks like you just move the `if (comm != MPI_COMM_NULL){}` block into the above `if---else if---`. Why do we do it?

**Review by minsii — CHANGES_REQUESTED** [review]
The two binding commits need to be divided into more commits (see details in my inlined comments). The current shape does not allow me to understand some logic of the code changes. I will make another round of review for the two binding commits after commit separation.

File: src/binding/custom_mapping.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+LIS_KIND_MAP:
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/custom_mapping.txt:1]: I can add notes to `doc/notes/binding_c.txt`. I also can add a short comment in `src/binding/custom_mapping.txt`. But let me also explain it here:

MPI forum repo uses "kind" to categorize each parameter, names such as `COMMUNICATOR` and `POLYXFER_NUM_ELEM_NNI`. Then it uses mapping, one for each language binding, to map the "kind" to "type". For example, `COMMUNICATOR` is mapped to `MPI_Comm`, and `POLYXFER_NUM_ELEM_NNI` is mapped to `int` or `MPI_Count` depending on embiggenning flavor.

There are a few `MPIX_` functions with mpich-specific parameters, such as "general request class" and "gpu device type". Because they are not in the standard, so we have to supply our own "KIND" and "mapping". 

The kind is used following each parameter in `mpi_standard_api.txt` or in our config files such as `src/binding/c/request_api.txt` for `MPIX_` functions. The mapping is in `src/binding/apis_mapping.txt` for standard upstream mapping, and in `custom_mapping.txt` for custom mappings (for mpix parameters)

File: src/include/mpir_misc.h (hunk 4 above)
Comment (hzhou) [reply src/include/mpir_misc.h:94]: They are normalized to e.g. `MPIR_Comm_join_impl` and prototypes are automatically generated by the python scripts in `src/include/mpir_impl.h`.

File: src/mpi/spawn/spawn_impl.c (hunk 5 above)
Comment (hzhou) [reply src/mpi/spawn/spawn_impl.c:82]: Will do.

File: src/mpi/spawn/spawn_impl.c (hunk 6 above)
Comment (hzhou) [reply src/mpi/spawn/spawn_impl.c:126]: All `MPIR/MPID` functions will use handle pointers (except for a few corner cases that we have to hard code the rules into the script). Remove the corner cases unless we have necessary reasons reduces the complexity that we have to maintain.

**Review by hzhou — COMMENTED** [review]
I will follow up with the split of commits.

File: src/binding/custom_mapping.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/custom_mapping.txt:4]: Comments and notes added to both `doc/notes/binding_c.txt` and `src/binding/custom_mapping.txt`

File: src/mpi/request/request_impl.c (hunk 1 above)
Comment (hzhou) [reply src/mpi/request/request_impl.c:10]: done

File: src/mpi/request/request_impl.c (hunk 3 above)
Comment (hzhou) [reply src/mpi/request/request_impl.c:99]: done.

File: src/mpi/spawn/spawn_impl.c (hunk 5 above)
Comment (hzhou) [reply src/mpi/spawn/spawn_impl.c:82]: done

File: test/mpi/errors/spawn/connect_timeout.c (hunk 8 above)
Comment (hzhou) [reply test/mpi/errors/spawn/connect_timeout.c:134]: There are two `MPI_Comm_connect`, one of them will succeed and the other one will return `MPI_COMM_NULL`. If the first one succeed, the second one (failed ones) will set comm to `MPI_COMM_NULL` and the comm objects from the first successful one will get leaked.

Since we don't know which `MPI_Comm_connect` will succeed, we need test each and free the one that succeeds (when `comm != MPI_COMM_NULL`).

File: test/mpi/errors/spawn/connect_timeout.c (hunk 7 above)
Comment (hzhou) [reply test/mpi/errors/spawn/connect_timeout.c:110]: I updated the fix by setting `comm = MPI_COMM_NULL` before each call. We are not relying the failure behavior but relying on that a **successful** `MPI_Comm_{connect,accept}` will set `comm` to a *valid* handle.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
@minsii Ready for review.

File: test/mpi/errors/spawn/connect_timeout.c (hunk 7 above)
Comment (minsii) [reply test/mpi/errors/spawn/connect_timeout.c:110]: What if a failed `MPI_Comm_{connect,accept}` set `comm` to a garbage value?  I think the test itself should follow the standard, i.e., free comm only when the return errno is `MPI_SUCCESS`?

Comment (hzhou) [reply test/mpi/errors/spawn/connect_timeout.c:110]: > What if a failed `MPI_Comm_{connect,accept}` set `comm` to a garbage value? I think the test itself should follow the standard, i.e., free comm only when the return errno is `MPI_SUCCESS`?

I see. That makes sense. Will update

File: src/mpi/request/request_impl.c (hunk 2 above)
Comment (hzhou) [reply src/mpi/request/request_impl.c:96]: fixed.

File: test/mpi/errors/spawn/connect_timeout.c (hunk 7 above)
Comment (hzhou) [reply test/mpi/errors/spawn/connect_timeout.c:109]: Fixed.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by minsii — CHANGES_REQUESTED** [review]
left some comments mainly for mpir_ refactoring. now checking autogen files.

File: src/binding/c/spawn_api.txt [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    + ip_port - Value contains IP port number at which to establish a port.
+    - ip_address - Value contains IP address at which to establish a port.
+     If the address is not a valid IP address of the host on which the
+     'MPI_OPEN_PORT' call is made, the results are undefined.
+*/
+
+MPI_Close_port:
+    .desc: close port
+    .docnotes: NotThreadSafe
+
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:70]: Is `MPI_Close_port ` not thread-safe defined in standard? the original file has 
```
MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
...
MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```

Comment (minsii) [reply src/binding/c/spawn_api.txt:70]: The generated file does not do line break format. but i think your rma pr should fix it for all files.

File: src/binding/c/spawn_api.txt [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+MPI_Comm_accept:
+    .desc: Accept a request to form a new intercommunicator
+
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:73]: comm_accept does not have `MPI_ERR_INFO`. same for comm_connect

File: src/binding/c/spawn_api.txt [hunk 12]
```diff
@@ -0,0 +1,106 @@
+# vim: set ft=c:
+
+MPI_Comm_get_parent:
+    .desc: Return the parent communicator for this process
+    .docnotes: SignalSafe
+{
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:6]: notes of `comm_get_parent` is lost.

File: src/binding/c/spawn_api.txt [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    *parent = (MPIR_Process.comm_parent == NULL) ? MPI_COMM_NULL :
+        (MPIR_Process.comm_parent)->handle;
+}
+
+MPI_Comm_join:
+    .desc: Create a communicator by joining two processes connected by a socket
+
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:13]: note of comm_join is lost

File: src/binding/c/spawn_api.txt [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+MPI_Comm_spawn_multiple:
+    .desc: Spawn up multiple MPI applications
+    .skip: ThreadSafe
+    .extra: errtest_comm_intra
+    .skip: validate-STRING_ARRAY, validate-STRING_2DARRAY, validate-COMM_SIZE, validate-INFO, validate-ERROR_CODE
+{ -- error_check --
+    if (comm_ptr->rank == root) {
+        MPIR_ERRTEST_ARGNULL(array_of_info, "array_of_info", mpi_errno);
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:39]: in spawn_multiple, check of each  info pointer is lost
```
    if (comm_ptr->rank == root) {
        MPIR_CHKLMEM_MALLOC(array_of_info_ptrs, MPIR_Info **, count * sizeof(MPIR_Info *),
                            mpi_errno, "array of info pointers", MPL_MEM_BUFFER);
        for (i = 0; i < count; i++) {
            MPIR_Info_get_ptr(array_of_info[i], array_of_info_ptrs[i]);
        }
    }
```

Comment (minsii) [reply src/binding/c/spawn_api.txt:39]: besides, when publishing the handle, the `if (intercomm_ptr) ` check seems unnecessary since a failed case will never go here
```
    mpi_errno = MPIR_Comm_spawn_multiple_impl(count, array_of_commands, array_of_argv, array_of_maxprocs, array_of_info, root, comm_ptr, &intercomm_ptr, array_of_errcodes);
    if (mpi_errno) {
        goto fn_fail;
    }
    if (intercomm_ptr) {
        MPIR_OBJ_PUBLISH_HANDLE(*intercomm, intercomm_ptr->handle);
    }
```

same for `comm_join`, `comm_spawn`

Comment (minsii) [reply src/binding/c/spawn_api.txt:39]: `MPI_ERR_SPAWN` is lost in comm_spawn*

File: src/binding/c/spawn_api.txt [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+/*
+    Notes:
+    This routine waits for all pending communication to complete, then frees the
+    communicator and sets 'comm' to 'MPI_COMM_NULL'.  It may not be called
+    with 'MPI_COMM_WORLD' or 'MPI_COMM_SELF'.
+*/
+
+MPI_Lookup_name:
+    .desc: Lookup a port given a service name
+/*
```
Comment (minsii) [inline src/binding/c/spawn_api.txt:90]: `MPI_ERR_INFO` is lost in lookup_name, publish_name

**Review by minsii — CHANGES_REQUESTED** [review]
comments for spawn autogen.

File: src/binding/c/request_api.txt [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+    elements associated with requests that have successfully completed will be
+    'MPI_SUCCESS'.  Finally, those requests that have not completed will have a
+    value of 'MPI_ERR_PENDING'.
+
+    While it is possible to list a request handle more than once in the
+    array_of_requests, such an action is considered erroneous and may cause the
+    program to unexecpectedly terminate or produce incorrect results.
+*/
+{ -- error_check -- array_of_statuses
+    if (count > 0) {
```
Comment (minsii) [inline src/binding/c/request_api.txt:402]: in waitall, `MPIR_ERRTEST_COUNT(count, mpi_errno);` is lost.

Comment (minsii) [reply src/binding/c/request_api.txt:402]: somehow `MPI_ERR_IN_STATUS` is lost in request files.

File: src/binding/c/request_api.txt [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+{
+    mpi_errno = MPIR_Waitall(count, array_of_requests, array_of_statuses);
+    if (mpi_errno)
+        goto fn_fail;
+}
+
+MPI_Waitany:
+    .desc: Waits for any specified MPI Request to complete
+    .docnotes: waitstatus
+/*
```
Comment (minsii) [inline src/binding/c/request_api.txt:415]: in waitany, `comm_ptr` is declared but is not set to request_ptr->comm

**Review by minsii — CHANGES_REQUESTED** [review]

File: src/mpi/request/greq_start.c [hunk 18]
```diff
@@ -266,16 +273,17 @@ int MPIX_Grequest_start(MPI_Grequest_query_function * query_fn,
...
     /* ... body of routine ... */
-    *request = MPI_REQUEST_NULL;
-    mpi_errno =
-        MPIX_Grequest_start_impl(query_fn, free_fn, cancel_fn, poll_fn, wait_fn, extra_state,
-                                 &lrequest_ptr);
-
-    if (mpi_errno == MPI_SUCCESS) {
-        *request = lrequest_ptr->handle;
+    MPIR_Request *request_ptr = NULL;
+    mpi_errno = MPIR_Grequest_start_impl(query_fn, free_fn, cancel_fn, extra_state, &request_ptr);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/request/greq_start.c]: Let's deleted line 271, it is a leftover.

File: src/mpi/request/greq_complete.c [hunk 19]
```diff
@@ -86,7 +86,10 @@ int MPI_Grequest_complete(MPI_Request request)
 
     /* ... body of routine ...  */
 
-    MPIR_Grequest_complete(request_ptr);
+    mpi_errno = MPIR_Grequest_complete_impl(request_ptr);
+    if (mpi_errno) {
+        return fn_fail;
+    }
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/request/greq_complete.c]: Fixed the `HAVE_ERROR_CHECKING` to only surround `MPIR_Err_create_code(...)`, so now it fits the pattern with the other mpi functions. Labels `fn_exit` and `fn_fail` are always in use now.

File: src/mpi/request/greq_start.c [hunk 20]
```diff
@@ -266,16 +273,17 @@ int MPIX_Grequest_start(MPI_Grequest_query_function * query_fn,
...
+    if (mpi_errno) {
+        goto fn_fail;
     }
+    request_ptr->u.ureq.greq_fns->poll_fn = poll_fn;
+    request_ptr->u.ureq.greq_fns->wait_fn = wait_fn;
+    *request = request_ptr->handle;
 
     /* ... end of body of routine ... */
+
     MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/request/greq_start.c]: fixed.

File: src/include/mpir_request.h [hunk 21]
```diff
@@ -564,24 +567,19 @@ int MPIR_Grequest_cancel(MPIR_Request * request_ptr, int complete);
...
+int MPIR_Grequest_class_create_impl(MPI_Grequest_query_function * query_fn,
                                     MPI_Grequest_free_function * free_fn,
                                     MPI_Grequest_cancel_function * cancel_fn,
                                     MPIX_Grequest_poll_function * poll_fn,
                                     MPIX_Grequest_wait_function * wait_fn,
                                     MPIX_Grequest_class * greq_class);
-int MPIX_Grequest_class_allocate_impl(MPIX_Grequest_class greq_class,
-                                      void *extra_state, MPI_Request * request);
+int MPIR_Grequest_class_allocate_impl(MPIX_Grequest_class greq_class,
+                                      void *extra_state, MPIR_Request ** p_request_ptr);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h]: The parameter type is changed so use a name to reflect the type. `p_request_ptr` means "pointer to a request pointer".

File: src/mpi/request/greq_start.c [hunk 22]
```diff
@@ -266,16 +273,17 @@ int MPIX_Grequest_start(MPI_Grequest_query_function * query_fn,
...
-    if (mpi_errno == MPI_SUCCESS) {
-        *request = lrequest_ptr->handle;
+    MPIR_Request *request_ptr = NULL;
+    mpi_errno = MPIR_Grequest_start_impl(query_fn, free_fn, cancel_fn, extra_state, &request_ptr);
+    if (mpi_errno) {
+        goto fn_fail;
     }
+    request_ptr->u.ureq.greq_fns->poll_fn = poll_fn;
+    request_ptr->u.ureq.greq_fns->wait_fn = wait_fn;
+    *request = request_ptr->handle;
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/request/greq_start.c]: > 
> 
> what is this change? It is not simple rename.
> 
> ```
>     request_ptr->u.ureq.greq_fns->poll_fn = poll_fn;
>     request_ptr->u.ureq.greq_fns->wait_fn = wait_fn;
> ```

`MPIX_Grequest_start` is nearly the same as `MPI_Grequest_start` except setting two additional function pointers (`poll_fn` and `wait_fn`). Rather than duplicate the code, the change now reuses `MPIR_Grequest_start_impl` (in stead of `MPIX_Grequest_start_impl`). The above two new lines are the difference that `MPIX_Grequest_start` needed.

The other reason is we have used the convention of naming MPIR impl functions as `MPIR_Xxx_impl` regardless it is for a `MPI_Xxx` function or a `MPIX_Xxx` function. Since here we have both `MPI_Grequest_start` and `MPIX_Grequest_start`. Let `MPIX_Grequest_start` call `MPIR_Grequest_start_impl` removes the need to create an exception for `MPIX_Grequest_start`.

> Besides, looks like `fn_fail` is not defined in this commit.

Thanks. Labels added.

File: src/mpi/request/request_impl.c [hunk 23]
```diff
@@ -407,9 +544,7 @@ int MPIR_Testany_state(int count, MPIR_Request * request_ptrs[],
     int mpi_errno = MPI_SUCCESS;
 
     mpi_errno = MPID_Progress_test(state);
-    /* --BEGIN ERROR HANDLING-- */
     MPIR_ERR_CHECK(mpi_errno);
-    /* --END ERROR HANDLING-- */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/request/request_impl.c:548]: It's just a comment and it seems out-of-place. Anyway, I put them back.

File: src/include/mpir_request.h [hunk 24]
```diff
@@ -567,20 +567,6 @@ int MPIR_Grequest_cancel(MPIR_Request * request_ptr, int complete);
...
-int MPIR_Grequest_class_create_impl(MPI_Grequest_query_function * query_fn,
-                                    MPI_Grequest_free_function * free_fn,
-                                    MPI_Grequest_cancel_function * cancel_fn,
-                                    MPIX_Grequest_poll_function * poll_fn,
-                                    MPIX_Grequest_wait_function * wait_fn,
-                                    MPIX_Grequest_class * greq_class);
-int MPIR_Grequest_class_allocate_impl(MPIX_Grequest_class greq_class,
-                                      void *extra_state, MPIR_Request ** p_request_ptr);
-
 /* These routines below are helpers for the Extended generalized requests. */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpir_request.h:570]: They are not removed. They are renamed to use the consistent pattern of MPIR_Xxx_impl. MPIX_Grequest_start_impl is removed since the name doesn't fit into MPIR_Xxx_impl , thus can be confusing. Instead we will have MPIX_Grequest_start call MPIR_Grequest_start_impl and set the additional poll_fn and wait_fn as we mentioned in the other comment.

File: src/binding/c/spawn_api.txt (hunk 12 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:6]: Notes added

Comment (hzhou) [reply src/binding/c/spawn_api.txt:6]: Also the `SignalSafe` docnotes is removed. We have discussed and agreed that MPI functions should not be declared "SignalSafe" as they are not part of the standard and it is easy to neglect the fact that libc signal unsafe functions is being used somewhere.

File: src/binding/c/spawn_api.txt (hunk 13 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:13]: Added

File: src/binding/c/spawn_api.txt (hunk 10 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:70]: > 
> 
> Is `MPI_Close_port ` not thread-safe defined in standard? the original file has
> 
> ```
> MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
> ...
> MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
> ```

I think it was a mistake to note `NotThreadSafe`. Let me remove it.

... removed.

File: src/binding/c/spawn_api.txt (hunk 11 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:73]: `MPIR_ERRTEST_INFO` and `MPIR_ERRTEST_INFO_OR_NULL` were using `MPI_ERR_ARG`. I added a new commit to fix it.

File: src/binding/c/spawn_api.txt (hunk 14 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:39]: > 
> 
> besides, when publishing the handle, the `if (intercomm_ptr) ` check seems unnecessary since a failed case will never go here
> 
> ```
>     mpi_errno = MPIR_Comm_spawn_multiple_impl(count, array_of_commands, array_of_argv, array_of_maxprocs, array_of_info, root, comm_ptr, &intercomm_ptr, array_of_errcodes);
>     if (mpi_errno) {
>         goto fn_fail;
>     }
>     if (intercomm_ptr) {
>         MPIR_OBJ_PUBLISH_HANDLE(*intercomm, intercomm_ptr->handle);
>     }
> ```
> 
> same for `comm_join`, `comm_spawn`

Probably, but I don't think it is critical. There are some places we need check for NULL pointers when returning handle objects. Always check for null pointers is simpler for the python script rather than create obscure rules for it.

Comment (hzhou) [reply src/binding/c/spawn_api.txt:39]: > 
> 
> `MPI_ERR_SPAWN` is lost in comm_spawn*

I don't see anywhere we return this error code.

Comment (hzhou) [reply src/binding/c/spawn_api.txt:39]: > 
> 
> in spawn_multiple, check of each info pointer is lost
> 
> ```
>     if (comm_ptr->rank == root) {
>         MPIR_CHKLMEM_MALLOC(array_of_info_ptrs, MPIR_Info **, count * sizeof(MPIR_Info *),
>                             mpi_errno, "array of info pointers", MPL_MEM_BUFFER);
>         for (i = 0; i < count; i++) {
>             MPIR_Info_get_ptr(array_of_info[i], array_of_info_ptrs[i]);
>         }
>     }
> ```

The allocation and populating of `array_of_info_ptrs` was moved to `spawn_impl.c` because there is no facility for the python scripts to handle info pointers array. I realize this makes `MPI_Comm_spawn` calling `MPIR_Comm_spawn_multiple` difficult since the former uses `info_ptr`. So I pushed a new commit to update python script to accept custom clean_up routine (which is needed for custom info_ptrs allocation), and have made corresponding changes in `spawn_api.txt`. Now the allocation of `array_of_info_ptrs` is in the custom handle_ptr routine inside `spawn_api.txt`

File: src/binding/c/spawn_api.txt (hunk 15 above)
Comment (hzhou) [reply src/binding/c/spawn_api.txt:90]: Fixed by the python update commit.

File: src/binding/c/request_api.txt [hunk 25]
```diff
@@ -0,0 +1,617 @@
+# vim: set ft=c:
+
+MPI_Cancel:
+    .desc: Cancels a communication request
+    .docnotes: NULL
+/*
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:6]: This is fixed from one of the update in the rma PR.

File: src/binding/c/request_api.txt [hunk 26]
```diff
@@ -0,0 +1,617 @@
...
+    Users are
+    advised that cancelling a send, while a local operation (as defined by the MPI
+    standard), is likely to be expensive (usually generating one or more internal
+    messages).
+*/
+
+MPI_Grequest_complete:
+    .desc: Notify MPI that a user-defined request is complete
+    .seealso: MPI_Grequest_start
+
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:25]: Added custom "error_check" routine with error code `MPI_ERR_REQUEST`.

File: src/binding/c/request_api.txt [hunk 27]
```diff
@@ -0,0 +1,617 @@
...
+            status->MPI_ERROR = MPI_SUCCESS;
+        }
+        goto fn_exit;
+    }
+}
+
+MPI_Start:
+    .desc: Initiates a communication with a persistent request handle
+{
+    mpi_errno = MPID_Startall(1, &request_ptr);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:104]: Added commit to add the "persistent" check to `MPI_Start`.
The `ARGNULL` check for "inout" handle point  has been fixed by a separate "python" commit.

File: src/binding/c/request_api.txt [hunk 28]
```diff
@@ -0,0 +1,617 @@
...
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+}
+
+MPI_Startall:
+    .desc: Starts a collection of persistent requests
+    .impl: mpid
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:111]: It was checked with `MPIR_ERRTEST_ARGNEG`, which essentially the same `MPIR_ERRTEST_COUNT`. Pushed a commit to use `MPIR_ERRTEST_COUNT` in stead.

File: src/binding/c/request_api.txt [hunk 29]
```diff
@@ -0,0 +1,617 @@
...
+    .docnotes: SignalSafe
+{
+    *flag = MPIR_STATUS_GET_CANCEL_BIT(*status);
+}
+
+MPI_Test:
+    .desc: Test for the completion of a request
+    .docnotes: waitstatus
+{
+    mpi_errno = MPIR_Test(request, flag, status);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:138]: > In test, `MPIR_ERRTEST_COUNT(count, mpi_errno);` is lost,

Did you mean in e.g. `MPI_Test{any,all,some}`? There is no count in `MPI_Test`. Anyway, this should be fixed with https://github.com/pmodels/mpich/pull/4967#discussion_r548780055

> and no `MPI_ERR_REQUEST`.

This is fixed with the "inout" handle python commit.

> maybe better to add back the note for temporary comm_ptr, since it is not straightforward until see fn_fail.

Added a commit to update python script to generate this comment.

File: src/binding/c/request_api.txt (hunk 17 above)
Comment (hzhou) [reply src/binding/c/request_api.txt:415]: It was an error to declare `comm_ptr` in the case of request array such as in `MPI_Waitany` as it is not clear which communicator from which request should be used.

The leaked `comm_ptr` declare is now fixed.

File: src/binding/c/request_api.txt (hunk 16 above)
Comment (hzhou) [reply src/binding/c/request_api.txt:402]: both are fixed.

File: src/binding/c/request_api.txt [hunk 30]
```diff
@@ -0,0 +1,617 @@
...
+    While it is possible to list a request handle more than once in the
+    'array_of_requests', such an action is considered erroneous and may cause the
+    program to unexecpectedly terminate or produce incorrect results.
+*/
+{ -- error_check -- array_of_statuses
+    if (count > 0) {
+        MPIR_ERRTEST_ARGNULL(array_of_statuses, "array_of_statuses", mpi_errno);
+    }
+}
+{
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/binding/c/request_api.txt:168]: It is wrong to allocate `request_ptrs` for `MPI_{Wait,Test}all` since they are handled inside `MPIR_{Wait,Test}`. fixed.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by minsii — APPROVED** [review]
I have read all responses and assume that all fixes have been properly in. I didn't re-generate & check these auto-generated files.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
